akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala

```diff
@@ -191,52 +192,79 @@ object Framing {
     val maximumFrameLength: Int,
-    val byteOrder: ByteOrder) extends PushPullStage[ByteString, ByteString] {
-    private var buffer = ByteString.empty
-    private var frameSize = Int.MaxValue
+    val byteOrder: ByteOrder) extends GraphStage[FlowShape[ByteString, ByteString]] {
     private val minimumChunkSize = lengthFieldOffset + lengthFieldLength
     private val intDecoder = byteOrder match {
       case ByteOrder.BIG_ENDIAN    ⇒ bigEndianDecoder
       case ByteOrder.LITTLE_ENDIAN ⇒ littleEndianDecoder
     }
 
-    private def tryPull(ctx: Context[ByteString]): SyncDirective =
-      if (ctx.isFinishing) ctx.fail(new FramingException("Stream finished but there was a truncated final frame in the buffer"))
-      else ctx.pull()
-
-    override def onPush(chunk: ByteString, ctx: Context[ByteString]): SyncDirective = {
-      buffer ++= chunk
-      doParse(ctx)
-    }
-
-    override def onPull(ctx: Context[ByteString]): SyncDirective = doParse(ctx)
+    val in = Inlet[ByteString]("LengthFieldFramingStage.in")
+    val out = Outlet[ByteString]("LengthFieldFramingStage.out")
+    override val shape: FlowShape[ByteString, ByteString] = FlowShape(in, out)
 
-    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective =
-      if (buffer.nonEmpty) ctx.absorbTermination()
-      else ctx.finish()
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var buffer = ByteString.empty
+      private var frameSize = Int.MaxValue
 
-    private def doParse(ctx: Context[ByteString]): SyncDirective = {
-      def emitFrame(ctx: Context[ByteString]): SyncDirective = {
-        val parsedFrame = buffer.take(frameSize).compact
+      /**
+       * push, and reset frameSize and buffer
+       *
+       */
+      private def pushFrame() = {
+        val emit = buffer.take(frameSize).compact
         buffer = buffer.drop(frameSize)
         frameSize = Int.MaxValue
-        if (ctx.isFinishing && buffer.isEmpty) ctx.pushAndFinish(parsedFrame)
-        else ctx.push(parsedFrame)
+        push(out, emit)
       }
 
-      val bufSize = buffer.size
-      if (bufSize >= frameSize) emitFrame(ctx)
-      else if (bufSize >= minimumChunkSize) {
-        val parsedLength = intDecoder(buffer.iterator.drop(lengthFieldOffset), lengthFieldLength)
-        frameSize = parsedLength + minimumChunkSize
-        if (frameSize > maximumFrameLength)
-          ctx.fail(new FramingException(s"Maximum allowed frame size is $maximumFrameLength but decoded frame header reported size $frameSize"))
-        else if (bufSize >= frameSize) emitFrame(ctx)
-        else tryPull(ctx)
-      } else tryPull(ctx)
-    }
+      /**
+       * try to push downstream, returns true if push success
+       *
+       * @return
+       */
+      private def tryPushFrame(): Boolean = {
+        var result = false
+        if (buffer.size >= frameSize) {
+          pushFrame
```

Comment:
and then ofc use the `()` when you call it to make it clear that it is a side effecting op rather than a pure function call!